official/resnet/ctl/ctl_imagenet_main.py

```diff
@@ -0,0 +1,295 @@
+        optimizer.apply_gradients(zip(grads, model.trainable_variables))
+
+        training_accuracy.update_state(labels, logits)
+        return loss
+
+      if strategy:
+        per_replica_losses = strategy.experimental_run_v2(
+            step_fn, args=(train_ds_inputs,))
+        return strategy.reduce(tf.distribute.ReduceOp.SUM, per_replica_losses,
+                               axis=None)
+      else:
+        return step_fn(train_ds_inputs)
+
+    def test_step(test_ds_inputs):
+      """Evaluation StepFn."""
+      def step_fn(inputs):
+        images, labels = inputs
+        logits = model(images, training=False)
+        loss = tf.keras.losses.sparse_categorical_crossentropy(labels,
+                                                               logits)
+        loss = tf.reduce_sum(loss) * (1.0/ flags_obj.batch_size)
+        test_loss.update_state(loss)
+        test_accuracy.update_state(labels, logits)
+
+      if strategy:
+        strategy.experimental_run_v2(step_fn, args=(test_ds_inputs,))
+      else:
+        step_fn(test_ds_inputs)
+
+    if flags_obj.enable_function:
+      train_step = tf.function(train_step)
+      test_step = tf.function(test_step)
+
+    epoch_exp_per_sec = []
+    stats = {}
+    for epoch in range(train_epochs):
+      logging.info('Starting to run epoch: %s', epoch)
+      train_iterator = iter(train_ds)
+
+      step = 0
+      total_loss = 0.0
+      batch_exp_per_sec = []
+      for step in range(train_steps):
+        start_time = time.time()
+        learning_rate = compute_learning_rate(
+            epoch + 1 + (float(step) / steps_per_epoch))
+        optimizer.lr = learning_rate
+        start_time = time.time()
+        total_loss += train_step(next(train_iterator))
+        end_time = time.time()
+        elapsed_time = end_time - start_time
+        samples_per_sec = flags_obj.batch_size / elapsed_time
+        # We skip the first step for warmup purposes. We can
+        # add a flag for tuning this.
+        if step > 0:
+          batch_exp_per_sec.append(samples_per_sec)
+
+        step += 1
+      train_loss = total_loss / step
+      # calculate average examples per second for a given epoch
+      epoch_exp_per_sec.append(np.mean(batch_exp_per_sec))
+      logging.info('Learning rate at epoch %s is %s',
+                   epoch, optimizer.lr.numpy())
+      logging.info('Training loss: %s, accuracy: %s%%',
+                   train_loss.numpy(),
+                   training_accuracy.result().numpy())
+      logging.info(
+          "Training Metric: {'epoch':%d, 'examples_per_second': %f}" %
+          (epoch, epoch_exp_per_sec[epoch]))
+
+      # Store the last train loss and accuracy calculated
+      stats['train_loss'] = train_loss.numpy()
+      stats['train_acc'] = training_accuracy.result().numpy()
+      training_accuracy.reset_states()
+
+      if (not flags_obj.skip_eval and
+          epoch % flags_obj.epochs_between_eval == 0):
```

Comment:
this would eval after the first epoch, even if epochs_between_eval = N>1. Is that what we want? I dont know what Keras fit does, but maybe we want to do only after the first N epochs.

Reply:
yes, makes sense. Done.

Reply:
Hmm i think this is still slightly wrong. if epochs_between_eval = 3, you want to eval at epoch #2, #5 etc, assuming you start at 0. so what you need it `(epoch + 1) % epochs_between_eval == 0` (basically coz of the zero indexing of epoch index). would be good to confirm what keras does.

Reply:
Keras starts from epoch + 1 in its calculation of validation frequency. I've done the same here as well.
why do you think evaluating at 3, 6 is incorrect? We do 4 epochs the first time(only because we skip 0) but other than that what is the value in doing 2, 5? maybe I am missing something here.